### ISSUE TITLE
xds: Consistency while handling CDS updates and EDS

### ISSUE THREAD

**Issue by sschepens** [envoyproxy/envoy#13009]
When a cluster is created or updated envoy it enters warming phase and needs a related ClusterLoadAssignement response to fully initialize.
During Envoy startup phase envoy sends requests for those resources to the management server and so the management servers knows it has to respond those.
But, when updating a Cluster via CDS, no EDS re-request is sent to the management server and management server doesn't really know it should send a ClusterLoadAssignement for that Cluster, even if the resource hasn't really changed.
This berhavior introduces subtle bugs in management servers, in our case, our resource versioning scheme somehow included the cluster version, yesterday we introduced a change to remove that, and that unexpectedly broke our cluster updates, leaving some clusters without traffic.

This should probably be handled by envoy, currently go-control-plane and java-control-plane don't really handle this since it's kind of hard to induce this behavior of always pushing an EDS update for a Cluster even if the resource hasn't really changed, and if not using ADS, envoy is possibly connected to multiple management servers which increases the difficulty.

@htuch suggested that envoy should probably unsubscribe from EDS for the updated cluster and immediately subscribe again, since this is what's actually happening inside envoy, a new Cluster is being created and it wants to subscribe to the resources, and the old one wants to unsubscribe. Some more thought should be done on this idea and the consecuences of the old cluster being unsubscribed from resources.

Some discussion already happened on Slack, opening this issue to continue discussion.

This issue also applies to LDS updates with RDS and probably SRDS with RDS.

**Comment by htuch** [comment]
CC @envoyproxy/api-shepherds

**Comment by markdroth** [comment]
> But, when updating a Cluster via CDS, no EDS re-request is sent to the management server and management server doesn't really know it should send a ClusterLoadAssignement for that Cluster, even if the resource hasn't really changed.

Why would the management server need to send an update for the EDS resource that has not changed, just because the CDS resource that pointed to it has changed?  This seems like a pretty clear Envoy bug: the CDS and EDS resources should be tracked independently from the perspective of the xDS client code, so that when a new Cluster object is created, it can just pick up the existing data for the EDS resource it wants.

In gRPC, the pattern we use here is to separate out the transport protocol logic from the data model logic.  The transport protocol logic (which includes not just the wire communication but also the caching logic, since those two things are ultimately closely tied together) is implemented in a single XdsClient object.  Its API allows callers to start watches for a given resource of a given type.  When the first watch for a given resource is started, the XdsClient will send a request to the xDS server subscribing to that resource, and all updates from the xDS server will be sent to that watcher.  It's fine for multiple callers to be watching the same resource: if a watch is started for a resource that is already subscribed to (because some other watch was already started), then the XdsClient will immediately send the already-cached data to the new watcher.  When all watchers for a given resource are cancelled, the XdsClient will send a request to the xDS server unsubscribing from that resource.

One significant difference between Envoy and gRPC is that when a CDS resource is updated, gRPC can update the individual fields in the cluster instance rather than having to create a whole new cluster instance.  But even if we did need to create a whole new cluster instance, it still would not pose a problem, because we'd keep the old instance around while the new instance warmed up.  This would mean that when the new instance would start a watch for the EDS resource name, it would get the value that was already cached in the XdsClient, and the management server would not know or care that the cluster instance had been replaced.

> @htuch suggested that envoy should probably unsubscribe from EDS for the updated cluster and immediately subscribe again, since this is what's actually happening inside envoy, a new Cluster is being created and it wants to subscribe to the resources, and the old one wants to unsubscribe. Some more thought should be done on this idea and the consecuences of the old cluster being unsubscribed from resources.

This seems sub-optimal, both because it will cause unnecessary network traffic and because it will make warming the new cluster instance take longer.  I think it should be possible to avoid all of that by separating out the xDS client code into its own layer, as I described above.

That having been said, my concern here is less about Envoy having sub-optimal behavior and more about making sure that the protocol semantics are clear.  I think that the protocol itself should not require clients to re-subscribe in this case, because I'd rather not change gRPC, which already handles this case correctly, to do something less optimal.  And if Envoy does do the less optimal thing, I am concerned that it may introduce situations where management servers are built with Envoy-specific assumptions that make it hard for them to support gRPC later.

**Comment by htuch** [comment]
@markdroth I think this was a deliberate choice we made when designing CDS warming. The idea is that you might also be switching the endpoints when changing cluster config, e.g. upgrading from non-TLS to TLS cluster. You can see a pretty detailed discussion in https://github.com/envoyproxy/envoy/issues/5168. 

On clarifying protocol semantics, I think we have a PoR to move forward here with https://github.com/envoyproxy/envoy/issues/11299. I'll be putting together a SoW draft for this during the week. This is probably the best way to get a formal and testable statement of intended semantics in these corner cases.

**Comment by markdroth** [comment]
Reading the discussion in #5168, the problem there actually seems like another symptom of the fact that Envoy's implementation does not cleanly separate the transport protocol layer (which by necessity includes caching) from the data model layer (in this case, the cluster instance).  IMHO, these should be two different layers with a clear separation between them, so that implementation details in the data model layer (such as completely replacing a cluster instance whenever one field changes) do not affect the behavior of the transport protocol layer (such as caching).

To say that another way, I think that the existence of this issue is an indication that the existing solution to #5168 didn't really solve the problem.  I think that if Envoy changed to cleanly separate these two layers, it would not only solve the immediate problem but also prevent future problems caused by this same underlying architectural problem.

In fact, it looks to me like you actually suggested something similar to this in https://github.com/envoyproxy/envoy/issues/5168#issuecomment-459812020, and I don't see any follow-up discussion there saying why this should not work.  I see that you said that it seems odd to do this in the gRPC mux impl, which I can understand, but I don't see why we can't wrap a fuller XdsClient abstraction around the gRPC mux impl.

**Comment by sschepens** [comment]
@htuch if one were to make something like change TLS without downtime, we now have `transport_socket_matches` and also, couldn't something be made with `AggregateClusters`?

**Comment by markdroth** [comment]
BTW, if you're interested in what the XdsClient API looks like in gRPC, the relevant part is here:

https://github.com/grpc/grpc/blob/139895d2c8a0e52540dfba323970f905a2d29df6/src/core/ext/xds/xds_client.h#L96

In our implementation, there's a separate watch API for each type of resource, mainly because we can't depend on the protobuf library and didn't want to make upb part of this API.  But in Envoy, you could easily have a single watch API used for all resource types, simply by specifying the resource type as a parameter when you start the watch and having the data reported to the watcher be an `Any`.

**Comment by sschepens** [comment]
I'm with @markdroth here.

I think from a protocol standpoint it pretty crappy for mangement servers to have to resend a configuration change if nothing really changed. Furthermore, in Incremental, Envoy is subscribed to resource updates and should only get notifiend when those actually change.

In the case of updating clusters to TLS, if it's not doable with current mechanics, maybe we're in need of additional methods for transparently doing this sort of updates, but it would seem that `transport_socket_matches` and `AggregateCluster` are some of those mechanics.

**Comment by htuch** [comment]
As in the original CDS work, our constraints are not greenfield, to some extent we need to be backwards compatible with existing management servers. I don't think it would be particularly hard to add caching to Envoy's existing implementation, the question is really what is the correct behavior on the wire, given Envoy's existing behavior, xDS server implementations and our desired goals. We've been running with the "CDS updates implies warming, implies an EDS response" for some time. Can we design something that works for folks who are assuming that and migrate to a better model? Potentially, but I don't think we can sunset Envoy's cluster warming behavior overnight.

**Comment by markdroth** [comment]
If Envoy stops requiring management servers to proactively send an EDS response after a CDS update, then it's not harmful for management servers that continue to do that (the update will effectively be a no-op on the client if nothing has changed), but it means that new management servers don't need to send that unnecessary update.

**Comment by htuch** [comment]
One thing I'd like to point out, going all the way back to the top of this thread, is that in Envoy today, what is going on, is that a cluster is going away and a new one is arriving, which happens to share the same name. When you delete a resource, in CDS, you logically unsubscribe from the EDS resources. We _could_ special case this to talk about what happens over an update, but keep in mind we're adding in an extra corner case to deal with in implementations.

I'm not opposed to any direction we take here. IMHO, the most logical thing would be for subscriptions to actually references the subscribing resource instance (e.g. via UUID), so that you could independently unsubscribe and resubscribe. But I don't think it's worth building that out just for this.

**Comment by htuch** [comment]
> If Envoy stops requiring management servers to proactively send an EDS response after a CDS update, then it's not harmful for management servers that continue to do that (the update will effectively be a no-op on the client if nothing has changed), but it means that new management servers don't need to send that unnecessary update.

It's not as simple as this, since some Envoy users will be expecting the warming behavior and want to defer the cluster warming until endpoint update. See the TLS upgrade example, which is not a hypothetical but a valid use of Envoy's existing mechanisms.

**Comment by htuch** [comment]
Also worth keeping in mind, this is not some undocumented behavior that people happen to be relying on, we've been guaranteeing this for some time, see https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/cluster_manager.html?highlight=warming#cluster-warming

**Comment by markdroth** [comment]
> One thing I'd like to point out, going all the way back to the top of this thread, is that in Envoy today, what is going on, is that a cluster is going away and a new one is arriving, which happens to share the same name. When you delete a resource, in CDS, you logically unsubscribe from the EDS resources. We could special case this to talk about what happens over an update, but keep in mind we're adding in an extra corner case to deal with in implementations.

I think the right way to model this is to say that the implicit deletion of RDS and EDS resources happens at the transport protocol layer, not at the data model layer -- which makes sense if you think about it, because this deletion is a cache operation, and caching is dictated by the transport protocol, not the data model.  In other words, inside of the XdsClient, when we receive a CDS update that removes a CDS resource, we remove the cached EDS resource that was referred to by the removed CDS resource, and we notify any watchers of that EDS resource that the resource no longer exists.  It should *not* be tied to the deletion of the data model object (the cluster instance) at all.

What this means is that when the XdsClient sees an updated CDS resource that changes some field *other* than the EDS service name, it knows that the EDS resource is still referenced and is therefore not being deleted.  The fact that Envoy will create a new cluster instance and destroy the old one at the data model layer should be irrelevant to that decision.

(Also, as we've discussed separately, I think that the implicit deletion of the RDS and EDS resources is a very ugly layering violation, specifically because it does require the transport protocol layer to know something about the relationship between the resources, which is really a data model concern.  IMHO we should remove this behavior as part of moving to the incremental protocol variants, which will make the layering cleaner here.)

> It's not as simple as this, since some Envoy users will be expecting the warming behavior and want to defer the cluster warming until endpoint update. See the TLS upgrade example, which is not a hypothetical but a valid use of Envoy's existing mechanisms.
> 
> Also worth keeping in mind, this is not some undocumented behavior that people happen to be relying on, we've been guaranteeing this for some time, see https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/cluster_manager.html?highlight=warming#cluster-warming

Interesting.  I agree, that will definitely make it harder to get rid of this.

Hmmm...  If management servers expect clients to not use the existing EDS data after a CDS update, is it going to cause problems that gRPC will use that existing data?  What if a management server is intentionally sending a CDS update that requires drastically different EDS data?  For example, if the CDS update switches from plaintext to TLS, and the EDS update points to a new set of endpoints that speak TLS?

This behavior seems like a clear layering violation and is *extremely* difficult to reason about, I really don't want to implement it in gRPC.  I think this is going to cause problems if we don't clean this up.

**Comment by markdroth** [comment]
Note that if we can find a way to get rid of the existing behavior, then in cases where a management server does not want the new cluster instance to use the old EDS data, it can always use a new EDS service name in the updated CDS resource.

**Comment by htuch** [comment]
I see this less as a layering violation as an attempt to bolt on consistency to an eventually consistent model. What warming on EDS is currently doing is saying "we may know the old EDS, but we want another EDS that is temporally fresher than this older one". Which might not even be true in an eventually consistent environment (i.e. not ADS), your warming EDS might be some delayed EDS update of data older than the cluster.

We've discussed trying to explicitly add in a notion of version dependency in the past, this probably isn't the right time to add it, but if we did have that, then it would be a lot clearer what is going on.

I agree it's pretty broken and hard to deal with. 

Maybe one path forward here is to add explicit fields to `Cluster` to describe what warming behavior we want to see (with the default retaining existing semantics).

**Comment by fxposter** [comment]
@htuch @markdroth I am very far from what you both discuss about separating data model and transport model, but - from what I see in APIs perspective - it basically says that dependencies between resources are described in terms of names. This "warming" thing from APIs perspective look like an implementation detail of envoy that is simply not possible to describe in terms of what API provides: cluster has changed, it's name is the same, management server's state from endpoints perspective has not changed - no updates by default should be here.
Also, if we go a bit further - if you get multiple updates to the same cluster in short period of time - that would require the "warming" to happen on the last update, but when you receive EDS "warming response" - is it possible for envoy to know - to which exact update is this EDS "warming response" related?

**Comment by htuch** [comment]
@fxposter Envoy just cares that it sees an EDS update after the CDS update, not which cluster it relates to. I don't see this as an Envoy implementation detail, this is clearly specified in the docs which are part of what currently constitutes the API for folks building control planes.

**Comment by markdroth** [comment]
> Maybe one path forward here is to add explicit fields to `Cluster` to describe what warming behavior we want to see (with the default retaining existing semantics).

That might very well be a useful migration path.  We could do a phased migration where we initially leave the default alone but provide the field to request the new behavior, then change the default but allow it to be disabled via the field, and then finally remove the field completely.

I just want to make sure that we wind up in a plce where this kind of layering violation is not supported.  Otherwise, we will never have a clean separation between the transport protocol and the data model.

**Comment by htuch** [comment]
As discussed with @markdroth offline, this is an area of xDS that is super confusing when we consider the mix of SotW, delta, ADS and non-ADS, different consistency models. From the above thread, it's also apparent that the current behavior is not well understood. I think we should put together a uRFC https://github.com/cncf/udpa/tree/master/proposals to clarify what the existing behavior is and what we want the future behavior to be, without breaking backwards compatibility for existing use cases where the current behavior makes sense.

**Comment by fxposter** [comment]
@htuch I mean a bit different thing. If this feature is intended for envoy not using "partially constructed" clusters in this particular case - then if there are multiple updates to the same cluster - when you receive an update to corresponding endpoints - you still don't know which cluster state that update of endpoints is related. And yes - "warming" will succeed, but it may succeed with wrong state.

ie: if cluster states (for single particular cluster) are CS1, CS2, CS3... and corresponding "good" endpoint states are ES1, ES2, ES3... Then if you receive an almost simultaneous updates "-> CS2" and "->CS3" - CDS will require warmup for "CS3" state. And if afterwards envoy receives state "ES2" and later "ES3", I can guess the first "fully constructed state" will be (CS3 + ES2), which might be incorrect.

**Comment by htuch** [comment]
@fxposter yes, unless you linearize on ADS you can get various kinds of races happening. This is what I was referring to when the ADS vs. non-ADS case being also pretty confusing above.

**Comment by htuch** [comment]
Here's my concrete proposal (which should be turned into a uRFC):
1. We add a field to `Cluster` that says `dont_warm_on_eds_for_updates`. When this is set, Envoy inherits the existing cluster's hosts (I think @snowp mentioned in the original issue that this was plausible).
2. We clarify for future xDS that resource subscriptions for resource Y from resource X remain active across updates to X.

None of this will change existing behaviors of LDS, CDS, etc. since we will potentially break folks who are relying on the existing warming behaviors.

**Comment by markdroth** [comment]
I think that's a fine proposal, but I would also like to see a migration path for fully eliminating the problematic behavior.  If we don't define such a path, we'll never be rid of it.

**Comment by sschepens** [comment]
@htuch shouldn't this also consider `Listener`? since they also warmup

**Comment by mattklein123** [comment]
The proposal in https://github.com/envoyproxy/envoy/issues/13009#issuecomment-689140016 SGTM. We can discuss the exact field name, etc. in code review.

I agree overall that this is an incredibly confusing detail that has tripped a lot of people up and we should fix this.

> I think that's a fine proposal, but I would also like to see a migration path for fully eliminating the problematic behavior. If we don't define such a path, we'll never be rid of it.

On the Envoy side we can talk about potentially flipping the default for the field, but we would have to give a lot of warning, runtime revert support, etc. as we have a large number of management servers in the wild that might break.

**Comment by ramaraochavali** [comment]
FWIW,  https://github.com/envoyproxy/envoy/pull/6021  implemented 1 on the proposal above

**Comment by snowp** [comment]
This has always been a surprising thing so I think it would be great to address. Similarly to others I've solved this management server side by making the EDS version dependent on the CDS version.

Without going through all the historical context I recall there being a conceptual point about whether xDS resources are defined just by their name: if we already know the value of EDS for foo, when we get an update for cluster foo that defines an EDS dependency, is the EDS foo still good? I think with the advancements towards UDPA this is pretty clearly yes, but I think it was less obvious back then.

As for solving this I can see two variations of https://github.com/envoyproxy/envoy/issues/13009#issuecomment-689140016: 
1) Reuse the old EDS value by copying the `Host` objects used by the old cluster into the new one
2) Reuse the old EDS value by processing a `onConfigUpdate` (or something similar) with the old EDS proto vaue

As @ramaraochavali said, #6021 implemented a variation of this (1), and this has some great properties around allowing connection reuse between cluster updates (since the `Host` owns the connection pools). That said, 2) feels a bit more inline with the separation of transport and data model layer idea: upon processing a new CDS update, Envoy simply checks its xDS cache for the EDS resource that the new cluster is asking for and adds in the endpoints immediately. If the EDS resource is missing, cluster warming is deferred until an EDS response is available.

**Comment by markdroth** [comment]
@mattklein123 

> On the Envoy side we can talk about potentially flipping the default for the field, but we would have to give a lot of warning, runtime revert support, etc. as we have a large number of management servers in the wild that might break.

I understand the need to provide a graceful migration here.  I can live with it taking a while as long as we do have a plan to get there and are actively moving toward that goal.  That way, when people ask about why gRPC and Envoy behave differently here, I can point to that plan and tell people that gRPC is already doing what Envoy will eventually do, and they should adjust their management servers accordingly.

**Comment by mattklein123** [comment]
Regarding the @snowp comment (https://github.com/envoyproxy/envoy/issues/13009#issuecomment-689482632), I would much prefer we do (2) vs. (1). IMO this is more clear and also more akin to what @markdroth is talking about as xDS as a transport/caching layer. I also think that if we do (2), the TLS upgrade case "just works" since the hosts would get re-added, re-health checked, etc.

**Comment by htuch** [comment]
The one concern around building "caching" into the transport layer is that for endpoints, this could add a fair bit of memory overhead. On the plus side, it's only on the main thread and doesn't need to be multiplied by number of workers, but at O(10k) endpoints this could start to show up.

**Comment by mattklein123** [comment]
> The one concern around building "caching" into the transport layer is that for endpoints, this could add a fair bit of memory overhead. On the plus side, it's only on the main thread and doesn't need to be multiplied by number of workers, but at O(10k) endpoints this could start to show up.

Yeah this is fair, though I wonder with some config options we could allow the user to make their own trade-offs.

**Comment by markdroth** [comment]
It's not clear to me how this caching can really be optional.  The xDS protocol itself requires the client to keep state.  If we don't do that, we wind up violating the protocol, which is why the caching and the transport protocol are so deeply intertwined.

That having been said, there is no reason that the cached data needs to be in the form of the original protos.  If we convert the EDS resource into a set of Hosts, then maybe we could just cache those hosts?  And if we want to avoid making a separate copy of the hosts in each Cluster instance while we replace one with another, maybe we could do something like make the hosts ref-counted, so that we can hold a ref in both the XdsClient layer and in each Cluster instance that uses them.

**Comment by snowp** [comment]
I think there is some concern for possible bugs now and in the future due to reusing `Host` objects. Redoing the whole warming process (with a cached response) will ensure that the new hosts come up properly configured for the cluster, so there is some appeal to doing this. 

That said, if the overhead of adding a caching layer might be good enough motivation to attempt to do the lower overhead but riskier approach?

**Comment by mattklein123** [comment]
> That said, if the overhead of adding a caching layer might be good enough motivation to attempt to do the lower overhead but riskier approach?

I would rather we not prematurely optimize. Per @markdroth I think we could translate to a memory optimized structure if required and throw away the proto. I would rather we do the caching approach since it does seem like the most correct solution and we can optimize it later?

**Comment by ramaraochavali** [comment]
I agree with @htuch  on caching concern.

> though I wonder with some config options we could allow the user to make their own trade-offs.

 My vote here would be to not  drive this by user config as it is purely an implementation detail and we will have to maintain two implementations.

>  I would rather we do the caching approach since it does seem like the most correct solution and we can optimize it later?

I agree, Caching approach is the most correct, less riskier and more future proof (handles some use cases like TLS upgrade automatically) but we need to think about memory optimized structure for endpoints from start - because it can throw people off with increased memory for large deployments.

**Comment by mattklein123** [comment]
It just occurred to me that aren't we now storing the EDS protos for config_dump? So I think we already have the memory overhead issue? Should we start by just making sure we don't duplicate the data twice before optimizing further?

**Comment by htuch** [comment]
No, we rebuild EDS from host data, see https://github.com/envoyproxy/envoy/blob/419428a82945a1a2ab0ab48e054e76ee851c5450/source/server/admin/config_dump_handler.cc#L219. It would be higher fidelity to have these around in a cache.

**Comment by mattklein123** [comment]
OK, anyway, let's just consider the config_dump aspect when we implement so we don't duplicate data.

**Comment by mpuncel** [comment]
Should this new mode only apply to gRPC? I'm not familiar enough with the REST and file based implementations to know if it's relevant there as well, but I'd guess not since REST fetches on intervals, and a file lookup should be dependable/envoy initiated?

Will this end up on the Subscription API which I believe is shared between all of these implementations?

And finally I assume this should be implemented for both delta and SotW? Would we tolerate a period where it is implemented for one and not the other?

**Comment by htuch** [comment]
It's a good question @mpuncel. I think the principle that as long as there is at least on subscription, we retain the last delivered configuration and make this available to warming clusters can easily be mapped to REST. I.e. if the next fetch is in 10s, but a cluster is warming now and we know its endpoints, seems reasonable to make them available. As we scale xDS and use CDNs+REST, having this capability seems increasingly more useful.

**Comment by markdroth** [comment]
Note that it might make sense to tackle this, #10793, and #2943 at the same time by adding an XdsClient layer that centralizes the stream management and caching behavior in one place.

**Comment by howardjohn** [comment]
One area we are seeing this as well is around Envoy's reconnecting to a control plane.

The two flows in Istio currently are:
* start of connection, expect CDS request and send EDS when its request
* Spontaneous push, we will always send CDS+EDS

This handles this issue sufficiently.

However, what we are seeing is that if Envoy disconnects from the control plane partially through processing a EDS and/or CDS response (not sure exactly where), the first request to the control plan will be EDS. So the control plane must actually send EDS twice. There _is_ a request for the EDS again after the CDS response, but its easy to accidentally treat it as an ACK since its identical to the ACK.

**Comment by htuch** [comment]
I kind of feel this is WAI but willing to be persuaded otherwise. Basically, when Envoy reconnects I don't think there are any restrictions on xDS ordering. Should there be? CC @markdroth @adisuissa

**Comment by howardjohn** [comment]
I am willing to believe this is WAI as well :slightly_smiling_face: however, I was quite surprised when I did see this (first time every in years of working with Envoy) so thought it was worth at least mentioning here

**Comment by kyessenov** [comment]
@htuch At some point, the bug with EDS update warming CDS cluster needs to be addressed. It complicates every control plane since they have to keep track of CDS/EDS update consistency, and it also makes it impossible to run them separately.

**Comment by markdroth** [comment]
No, I don't think there should be any restrictions on what order subscriptions occur in after a stream restart (or any other time, for that matter).  I think the underlying problem here is still the same one described in this issue: it's that Envoy currently expects EDS to be resent after a CDS update.  If we fix that underlying problem, then this latest problem goes away, because it's just a side-effect of the underlying problem.

I believe that @adisuissa has started working on a design for an xDS caching layer in Envoy, as discussed earlier in this issue.

FWIW, note that there's a good description of how the xDS client code should be structured in https://github.com/cncf/xds/blob/main/proposals/TP2-dynamically-generated-cacheable-xds-resources.md#background-xds-client-and-server-architecture.  I've been meaning to lift that out of there and move it into the main xDS spec.

**Comment by htuch** [comment]
@kyessenov yes, these are separate issues I think and the warming issues needs to be addressed. Caching layer will generally help.

**Comment by adisuissa** [comment]
> However, what we are seeing is that if Envoy disconnects from the control plane partially through processing a EDS and/or CDS response (not sure exactly where), the first request to the control plan will be EDS. So the control plane must actually send EDS twice. There _is_ a request for the EDS again after the CDS response, but its easy to accidentally treat it as an ACK since its identical to the ACK.

@howardjohn I might be missing something, but is the `response_nonce` the same for the EDS request and the ACK?

I think we should add more tests in this area, clarify the expected behavior, and change if needed.

**Comment by howardjohn** [comment]
> > However, what we are seeing is that if Envoy disconnects from the control plane partially through processing a EDS and/or CDS response (not sure exactly where), the first request to the control plan will be EDS. So the control plane must actually send EDS twice. There _is_ a request for the EDS again after the CDS response, but its easy to accidentally treat it as an ACK since its identical to the ACK.
> 
> @howardjohn I might be missing something, but is the `response_nonce` the same for the EDS request and the ACK?
> 
> I think we should add more tests in this area, clarify the expected behavior, and change if needed.

Yes, there are two identical DiscoveryRequests (I would call these "ack" and "re-request EDS")

**Comment by hzxuzhonghu** [comment]
We suffered much from distinguishing **discovery request** or **ACK** in istio. Can envoy add an explicit flag to tell control plane about this?

**Comment by costinm** [comment]
+1 - having to guess if a message is an ACK or a request instead of having an explicit type is very uncommon in protocols and leads to a lot of bugs.  

It is also pretty easy and low risk to fix by adding a new field to indicate the message is an ACK: for NACK the presence of error_detail is used, but I don't think it is possible to determine if it is an ACK only by the content of the message.
A bool "ack" or int "type" would work ( "type" could also explicitly cover NACK, ACK, REQUEST, RECONNECT, etc).

**Comment by kyessenov** [comment]
In general, ACK and REQUEST are immediately called in succession in Envoy because of long poll. So even if there are two messages, it's the same problem.

**Comment by hzxuzhonghu** [comment]
@kyessenov I can not understand that, after `REQUEST `, istio or other control plane just need to send response. And we will not miss pushing a response.

**Comment by htuch** [comment]
This ACK/REQUEST discussion is fairly off topic - please open a new issue.

(FWIW, @kyessenov is right - even if we did add this feature, it wouldn't be materially different for Envoy, we would always send both ACK and REQUEST back-to-back. We did discuss making the protocol more explicit way back when we were looking at a breaking transport change for UDPA, but decided against that. There is some work being done by @markdroth right now around https://github.com/envoyproxy/envoy/issues/11396 and also other related issues at https://github.com/envoyproxy/envoy/issues/18887 and https://github.com/envoyproxy/envoy/issues/5739. We can do better here, but should structure this as part of a general improvement to the transport error feedback).

**Comment by costinm** [comment]
Confused, why would you send ACK and REQUEST back to back ? The protocol
has REQUEST, response and ACK.
The server needs to respond to a request - but not to an ack.

The root cause for this issue - and many others - is that it is pretty hard
to distinguish between the ack and request, since they have the exact
same content - which is pretty rare in protocols but also very easy to fix
- and would avoid a lot of complexity and bugs in servers.

**Comment by costinm** [comment]
This is not about making things materially different for envoy - it is
about making things cleaner on the servers, and following the patterns that
all protocols
follow. XDS is effectively a pubsub protocol, and I don't know of any
similar protocol that doesn't have a clear ACK message, or having identical
message mean different things. Alsoi seems to be the simplest/cleanest fix
to this particular issue ?

**Comment by markdroth** [comment]
FWIW, I agree that the semantics of ACK vs. request in the transport protocol could be made clearer, and I think that's something we should work on.  However, as @htuch said, I think that's a separate problem from the caching functionality that is the subject of this issue.  In this case, there should not be a subsequent request in the first place, because the client should not be re-requesting a resource that it is already subscribed to.  I think we should add caching to solve that problem rather than trying to hack in some way for the client to re-request something that it should already have.

**Comment by hzxuzhonghu** [comment]
Any way could prevent the hack judgement in istio 

https://github.com/istio/istio/blob/5f90e4b9ae19800f4c539628ae038ec118835610/pilot/pkg/xds/ads.go#L392-L406

**Comment by stevenzzzz** [comment]
cc @stevenzzzz

**Comment by lepistone** [comment]
Hi! We've been hit by this old issue on a custom SoTW control plane that serves both gRPC and Envoy.
I was wondering if there is suggested way (maybe from @markdroth) of how a control plane should handle cluster warming in the context of xDS Federation as described in [TP1](https://github.com/cncf/xds/blob/main/proposals/TP1-xds-transport-next.md) and TP2? The control planes for CDS and EDS can very well have different authorities and be deployed separately. So I wonder what's the ideal behavior for the control plane here. Implementing coordination between the two (especially to recognize the ACK-shaped EDS request which must be responded to) is more difficult across servers, especially when trying to achieve strong cacheability. In some cases it's an acceptable workaround to change the name of the ClusterLoadAssignment, but that's not always practical.
To make things worse, IIUC the EDS cache doesn't work when using federated xDS, since only the global adsMux is used for the cache.
Thank you!

**Comment by markdroth** [comment]
I agree, it's not reasonable to expect the control planes to handle this when CDS and EDS come from different control planes.

From an xDS perspective, the fact that Envoy needs to re-request the EDS resource after a CDS update is a bug in Envoy.  I think the fix is to give Envoy a true caching layer as described earlier in this issue.

I think there was broad agreement that Envoy needs this caching layer, but no one ever stepped up to implement it.  I would encourage someone to do so. :)
